core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -953,7 +953,7 @@ public boolean ignoreRecoveryAttempt() {
 
     public void readAllowed() throws IllegalIndexShardStateException {
         IndexShardState state = this.state; // one time volatile read
-        if (state != IndexShardState.STARTED && state != IndexShardState.RELOCATED) {
+        if (state != IndexShardState.STARTED && state != IndexShardState.RELOCATED && state != IndexShardState.POST_RECOVERY) {
             throw new IllegalIndexShardStateException(shardId, state, "operations only allowed when started/relocated");
```

Comment:
We should probably update the exception message to be `started/relocated/post-recovery` also?